pkg/pushsync/pushsync.go

```diff
@@ -332,118 +329,101 @@ func (ps *PushSync) pushToClosest(ctx context.Context, ch swarm.Chunk, origin bo
-					return swarm.ZeroAddress, false, ErrWarmup
+					return swarm.ZeroAddress, ErrWarmup
 				}
 
-				if skipPeers.OverdraftListEmpty() { // no peers in skip list means we can be confident that we are the closest peer
-					// we don't act on ErrWantSelf unless there are no overdraft peers
-					if !ps.radiusChecker.IsWithinStorageRadius(ch.Address()) {
-						return swarm.ZeroAddress, false, ErrOutOfDepthStoring
-					}
-					ps.pushToNeighbourhood(ctx, fullSkipList, ch, origin, originAddr)
-					return swarm.ZeroAddress, false, err
+				if !ps.radiusChecker.IsWithinStorageRadius(ch.Address()) {
+					return swarm.ZeroAddress, ErrOutOfDepthStoring
 				}
 
-				ps.logger.Debug("pushsync: continue iteration and reset overdraft skiplist")
-
-				skipPeers.ResetOverdraft() // reset the overdraft list and retry (in case the closest peer was there)
-				return swarm.ZeroAddress, true, nil
+				ps.pushToNeighbourhood(ctx, fullSkipList, ch, origin, originAddr)
+				return swarm.ZeroAddress, err
 			}
 
-			return swarm.ZeroAddress, false, fmt.Errorf("closest peer: %w", err)
+			return swarm.ZeroAddress, fmt.Errorf("closest peer: %w", err)
 		}
-
-		return peer, false, nil
+		return peer, nil
 	}
 
-	for {
+	for sentErrorsLeft > 0 {
 		select {
 		case <-ctx.Done():
 			return nil, ErrNoPush
-		case <-timer.C:
-
-			allowedRetries--
-			// decrement here to limit inflight requests, if the request is not "attempted", we will increment below
+		case <-preemptiveTicker:
+			retry()
+		case <-retryC:
 
-			peer, retry, err := nextPeer()
+			peer, err := nextPeer()
 			if err != nil {
-				return nil, err
-			}
-
-			if retry {
-				if allowedRetries <= 0 {
-					return nil, ErrNoPush
+				if inflight == 0 {
+					ps.logger.Debug("no peers left to retry", "chunk_address", ch)
+					return nil, fmt.Errorf("get closest for address %s, allow upstream %v: %w", ch, origin, err)
 				}
-				timer.Reset(waitRefresh)
 				continue
 			}
+			skip = append(skip, peer)
 
 			ps.metrics.TotalSendAttempts.Inc()
 
-			skipPeers.Add(peer)
-
-			allowedPushes--
+			inflight++
 
 			go func() {
 				ctxd, cancel := context.WithTimeout(ctx, defaultTTL)
 				defer cancel()
-				ps.pushPeer(ctxd, resultChan, doneChan, peer, ch, origin)
+				ps.pushPeer(ctxd, resultChan, done, peer, ch, origin)
 			}()
 
-			// reached the limit, do not set timer to retry
-			if allowedRetries <= 0 || allowedPushes <= 0 {
-				continue
-			}
-
-			// retry
-			timer.Reset(p90TTL)
-
 		case result := <-resultChan:
 
-			ps.measurePushPeer(result.pushTime, result.err, origin)
+			inflight--
 
-			if errors.Is(result.err, errNotAttempted) {
-				logger.Debug("not attempted: adding overdraft peer to skiplist", "peer_address", result.peer)
-				skipPeers.AddOverdraft(result.peer)
-			}
+			ps.measurePushPeer(result.pushTime, result.err, origin)
 
-			if ps.warmedUp() && !errors.Is(result.err, errNotAttempted) {
+			if ps.warmedUp() && !errors.Is(result.err, errPrepareCredit) {
```

Comment:
If we dont add these peers (errPrepareCredit) to the skiplist, the closest peer will keep returning the same peer first. This would cause us to retry prematurely. Previously we would have skipped and then cleared the overdraft list and only then consider this peer.

Reply:
these peers get added to the other skiplist already that is discarded after the request terminates.
The `ps.skiplist` is the global one.